crates/ide_assists/src/handlers/merge_match_arms.rs

```diff
@@ -88,6 +94,32 @@ fn contains_placeholder(a: &ast::MatchArm) -> bool {
     matches!(a.pat(), Some(ast::Pat::WildcardPat(..)))
 }
 
+fn are_same_types(
+    current_arm_types: &Vec<Option<hir::TypeInfo>>,
+    arm: &ast::MatchArm,
+    ctx: &AssistContext,
+) -> bool {
+    let arm_types = get_arm_types(&ctx, &arm);
+    for i in 0..arm_types.len() {
+        let other_arm_type = &arm_types[i];
+        let current_arm_type = &current_arm_types[i];
+        if let (Some(other_arm_type), Some(current_arm_type)) = (other_arm_type, current_arm_type) {
+            return &other_arm_type.original == &current_arm_type.original;
+        }
+    }
+
+    return true;
+}
+
+fn get_arm_types(ctx: &AssistContext, arm: &ast::MatchArm) -> Vec<Option<hir::TypeInfo>> {
+    match arm.pat() {
+        Some(ast::Pat::TupleStructPat(tp)) => {
```

Comment:
From my testing, it appeared as if `TupleStructPat` is the type of an arm -- is that correct?

Reply:
An arm can have any kind of pattern, not just a `TupleStructPat`. I think what we should do here is recursively walk the pattern of the current arm, collect and collect all `IdentPat` that have a `pat.name()`(function returns an `Option`) and do a `ctx.sema.resolve_pat(ast::Pat::IdentPat(ident_pat))` to get the type of the bound identifier.

So instead we should return something along the lines of a `FxHashMap<String, TypeInfo>` where String is the name of the ident pat and type info the corresponding type of it.

Then in the `arms_to_merge` iter we should check that all `IdentPat` names in those arms have a matching type with the current one by looking up their name in the hashmap and compare the types.

Reply:
I certainly can. Do you have an example of a match that doesn't use a `TupleStructPat`?

Reply:
I skipped a few chapters and took a look at Patterns. I've created this commit which adds a couple more unit tests for all sorts of different usages. Right now 6 unit tests are failing:

```
handlers::merge_match_arms::tests::merge_match_arms_enum_destructuring_different_types
handlers::merge_match_arms::tests::merge_match_arms_enum_destructuring_same_types_different_name
handlers::merge_match_arms::tests::merge_match_arms_enum_nested_pattern_different_names
handlers::merge_match_arms::tests::merge_match_arms_enum_without_field
handlers::merge_match_arms::tests::merge_match_arms_range
handlers::merge_match_arms::tests::merge_match_same_destructuring_different_types
```

I'll see what I can do to correctly handle these scenarios (right now all of these offer to generate invalid code).

Reply:
@Veykril Thanks for the suggestions! I've made some changes to follow your idea and all tests pass -- I believe this should give pretty good coverage of the different ways patterns can be used.

My instinct was to go for a `HashMap<IdentPat, Option<TypeInfo>>` but the `IdentPat` didn't seem to be working as I expected it to. I then switched that to `HashMap<String, Option<TypeInfo>>` by calling `.text().to_string()` and straight away all my tests passed.